tools/c7n_azure/c7n_azure/resources/key_vault_keys.py

```diff
@@ -0,0 +1,97 @@
+# Copyright 2018 Capital One Services, LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.from c7n_azure.provider import resources
+
+import logging
+
+from azure.keyvault.key_vault_id import KeyVaultId
+
+from c7n.filters import Filter
+from c7n.utils import type_schema
+
+from c7n_azure.provider import resources
+from c7n_azure.query import ChildResourceManager
+from c7n_azure.utils import ThreadHelper, ResourceIdParser
+
+
+log = logging.getLogger('custodian.azure.keyvault.keys')
+
+
+@resources.register('keyvault-keys')
+class KeyVaultKeys(ChildResourceManager):
+
+    class resource_type(ChildResourceManager.resource_type):
+        service = 'azure.keyvault'
+        client = 'KeyVaultClient'
+        enum_spec = (None, 'get_keys', {
+            'vault_base_url': lambda p: 'https://{0}.vault.azure.net'.format(p['name'])
+        })
+        parent_spec = ChildResourceManager.ParentSpec(
+            manager_name='keyvault',
+            annotate_parent=True
+        )
+
+
+@KeyVaultKeys.filter_registry.register('keyvault')
+class KeyvaultFilter(Filter):
+    schema = type_schema(
+        'keyvault',
+        required=['keyvaults'],
+        **{
+            'keyvaults': {'type': 'array', 'items': {'type': 'string'}}
+        }
+    )
+
+    def process(self, resources, event=None):
+        return [r for r in resources
+                if ResourceIdParser.get_resource_name(r['c7n:parent-id']) in self.data['keyvaults']]
+
+
+@KeyVaultKeys.filter_registry.register('key-type')
+class KeyTypeFilter(Filter):
+    schema = type_schema(
+        'key-type',
+        **{
+            'key-types': {'type': 'array', 'items': {'enum': ['EC', 'EC-HSM', 'RSA', 'RSA-HSM']}}
+        }
+    )
+
+    def process(self, resources, event=None):
+
+        resources, _ = ThreadHelper.execute_in_parallel(

```

Comment:
having spent most of the day debugging a concurrency issue... :-( would be nice to just define the other side of this aka `_` instead of making it anonymous.

Reply:
I think this needs to be case-by-case thing. If execute_in_parallel can return an exceptions list, for sure we have to capture it. However, in that particular case there is no possible exceptions returned (If we fail to query keys from a kv, we log warning and don't raise).

So, in this case, using anonymous means 'no exceptions possible'.